### PR TITLE
Add py3 native image

### PR DESCRIPTION
**Pull request by GuyAfik**
## Status
Ready

## Related Content Pull Request
Related PR: link to the PR at demisto/content

## Related Issues
[link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-4374)

## Description
add py3 master image

### CODE REVIEW

File: docker/py3-primary/chromium/download_chromedriver.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env bash
+
+# exit on errors
+set -e
+
+###########################
+# Script to download relevant chromedriver according to installed chrome version
```
Comment (JudahSchwartz) [inline docker/py3-primary/chromium/download_chromedriver.sh:7]: wdym installed chrome version? Can we use this for chromium image too?

File: docker/py3-primary/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+# will always install the latest stable google-chrome version
+RUN dnf update --nodocs -y  \
+    && dnf upgrade -y --nodocs \
+    && dnf install -y wget \
+    && wget https://dl.google.com/linux/linux_signing_key.pub \
+    && wget https://www.centos.org/keys/RPM-GPG-KEY-CentOS-Official \
+    && rpm --import linux_signing_key.pub \
+    && rpm --import RPM-GPG-KEY-CentOS-Official \
+    && dnf install -y --nodocs python3-devel gcc gcc-c++ make wget git unzip llvm-libs google-chrome-stable \
```
Comment (JudahSchwartz) [inline docker/py3-primary/Dockerfile:19]: google-chrome-stable will take from google-chrome in the repo folder?

File: docker/py3-primary/office-utils/LibreOffice_7.4.2.3_Linux_x86-64_rpm/readmes/README_en-US [hunk 3]
```diff
@@ -0,0 +1,202 @@
+
+======================================================================
+LibreOffice 7.4 ReadMe
```
[the remark this answers is not in the document]
Comment (GuyAfik) [reply docker/py3-primary/office-utils/LibreOffice_7.4.2.3_Linux_x86-64_rpm/readmes/README_en-US:3]: it was inside the tar, can be removed

File: docker/py3-primary/office-utils/LibreOffice_7.4.2.3_Linux_x86-64_rpm/install [hunk 4]
```diff
@@ -0,0 +1,319 @@
+#!/usr/bin/env bash
+#
+# This file is part of the LibreOffice project.
```
[the remark this answers is not in the document]
Comment (GuyAfik) [reply docker/py3-primary/office-utils/LibreOffice_7.4.2.3_Linux_x86-64_rpm/install:3]: this is the result of the zip extraction from the official libreoffice site: https://www.libreoffice.org/donate/dl/rpm-x86_64/7.4.2/en-US/LibreOffice_7.4.2_Linux_x86-64_rpm.tar.gz - i was unable to untar it inside the container

File: docker/py3-primary/chromium/download_chromedriver.sh (hunk 1 above)
Comment (GuyAfik) [reply docker/py3-primary/chromium/download_chromedriver.sh:7]: its taken directly from chromium image, its the same script.

it downloads the corresponding chromedriver according to the google-version installed

File: docker/py3-primary/Dockerfile (hunk 2 above)
Comment (GuyAfik) [reply docker/py3-primary/Dockerfile:19]: exactly

File: docker/py3-primary/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+packaging = "*"
+catboost = "*"
+rtfde = "*"
+webdriver-manager = "*"
+selenium = "*"
+cymruwhois = "*"
+simplejson = "*"
+networkx = "*"
+slackclient = "*"
+regex = "==2022.3.2"  # seems to be a bug in versions > 2022.3.2 https://stackoverflow.com/questions/71496687/dateparser-throws-regex-regex-core-error
```
[the remark this answers is not in the document]
Comment (GuyAfik) [reply docker/py3-primary/pyproject.toml:64]: dev-bug that its purpose is..?

File: docker/py3-primary/verify.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
+# chromium
+from selenium import webdriver
```
[the remark this answers is not in the document]
Comment (GuyAfik) [reply docker/py3-primary/verify.py:2]: 1. not all the images have verify.py, some of them are missing (like tesseract for example)
2. i believe this is a scope of another dev-task - this shouldn't be hard though

File: docker/py3-primary/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+# this is being done because when trying to unzip the rpm inside the container doesn't work, hence copying it as rpm.
+COPY office-utils/LibreOffice_7.4.2.3_Linux_x86-64_rpm LibreOffice_Linux_x86-64_rpm
+
+# handles os dependancies for: crypto + readpdf + parse-emails + docxpy + sklearn + pandas + ippysocks-py3 + oauthlib
+# unzip + py3ews + taxii2 + pan-os-python + slackv3 + google-api-py3 + boto3py3
+# pyjwt3 + joe-security + slack + mlurlphishing + office-utils
+# note: the packages that are being removed by dnf remove at the end are packages that are required during the installation of python dependancies
+
+# crypto os dependancies: python3-devel gcc gcc-c++ make wget git rust cargo libffi-devel openssl-devel
+# readpdf os dependancies: poppler poppler-utils
```
[the remark this answers is not in the document]
Comment (GuyAfik) [reply docker/py3-primary/Dockerfile:65]: i agree - this is just for now for making stuff clear during development

File: .github/dependabot.yml [hunk 8]
```diff
@@ -789,6 +789,10 @@ updates:
     directory: /docker/cymruwhois
     schedule:
      interval: daily
+  - package-ecosystem: pip
```
[the remark this answers is not in the document]
Comment (GuyAfik) [reply .github/dependabot.yml:792]: i don't find such option

File: docker/build_docker.sh [hunk 9]
```diff
@@ -170,6 +170,23 @@ function docker_build {
         # del_requirements=yes
     fi
 
+    if [ -f "pyproject.toml" -a ! -f "requirements.txt" ]; then
+       if [ ! -f "poetry.lock" ]; then
+            echo "Error: pyproject.toml present without poetry.lock. Make sure to commit your poetry.lock file"
+            return 1
+        fi
+
+      curl -sSL https://install.python-poetry.org | python3 -  # download poetry
```
[the remark this answers is not in the document]
Comment (GuyAfik) [reply docker/build_docker.sh:179]: do you mean to add it to the `dockerbuild` image? - https://github.com/demisto/dockerfiles/tree/master/docker/dockerbuild

File: .github/dependabot.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (GuyAfik) [reply .github/dependabot.yml:792]: just found it, will update to poetry https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#package-ecosystem

File: docker/py3-primary/chromium/download_chromedriver.sh [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+# for testing pruposes (outside of the container) you can set GOOGLE_CHROME_VERSION env var
+# for example: 
+# docker run --rm -it -v `pwd`:/work -w /work -e GOOGLE_CHROME_VERSION=91.0.4472 demisto/chromium:1.0.0.23161 docker/chromium/download_chromedriver.sh
+if [ -z "$GOOGLE_CHROME_VERSION" ]; then
+    echo "Determing chrome version..."
+    GOOGLE_CHROME_VERSION=$(google-chrome --version | grep -o -P '\d+\.\d+\.\d+')
+fi
+
+if [ -z "$GOOGLE_CHROME_VERSION" ]; then
+    echo "something went wrong with getting chrome version. Aborting!!!"
```
Comment (tkatzir) [inline docker/py3-primary/chromium/download_chromedriver.sh:19]:
```suggestion
    echo "Something went wrong with getting chrome version. Aborting!!!"
```

File: docker/py3-primary/chromium/download_chromedriver.sh [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    exit 1
+fi 
+
+echo "using GOOGLE_CHROME_VERSION: $GOOGLE_CHROME_VERSION"
```
Comment (tkatzir) [inline docker/py3-primary/chromium/download_chromedriver.sh:23]:
```suggestion
echo "Using GOOGLE_CHROME_VERSION: $GOOGLE_CHROME_VERSION"
```

File: docker/py3-primary/chromium/download_chromedriver.sh [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+echo "downloading chromedriver list..."
```
Comment (tkatzir) [inline docker/py3-primary/chromium/download_chromedriver.sh:25]:
```suggestion
echo "Downloading chromedriver list..."
```

File: docker/py3-primary/chromium/download_chromedriver.sh [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+wget -O chromedriver.list.xml  https://chromedriver.storage.googleapis.com/ 
+
+DRIVER_VERSION=$(grep -o -P "$GOOGLE_CHROME_VERSION\.\d+/chromedriver_linux64.zip" chromedriver.list.xml | sort -V | tail -1 | awk -F '/' '{print $1}')
+
+echo  "Using chromedriver version: $DRIVER_VERSION"
```
Comment (tkatzir) [inline docker/py3-primary/chromium/download_chromedriver.sh:30]:
```suggestion
echo "Using chromedriver version: $DRIVER_VERSION"
```

**Review by tkatzir — COMMENTED** [review]
Looks great!
Can we store the binary files somewhere outside the source tree?

File: .github/dependabot.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (GuyAfik) [reply .github/dependabot.yml:792]: getting this error: `The property '#/updates/177/package-ecosystem' value "poetry" did not match one of the following values: npm, bundler, composer, maven, mix, cargo, gradle, nuget, gomod, docker, elm, gitsubmodule, github-actions, pip, terraform, pub`

File: docker/py3-primary/office-utils/LibreOffice_7.4.2.3_Linux_x86-64_rpm/install (hunk 4 above)
[the remark this answers is not in the document]
Comment (GuyAfik) [reply docker/py3-primary/office-utils/LibreOffice_7.4.2.3_Linux_x86-64_rpm/install]: removed

File: docker/py3-primary/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+    && google-chrome --version \
+    && chromedriver --version \
+    && dnf remove -y python3-devel gcc gcc-c++ make platform-python-pip unzip \
+    && dnf clean all \
+    && rm -rf /var/cache/dnf \
+    && rm -rf download_chromedriver.sh
+
+
+# tesseract
+RUN dnf install -y git \
```
[the remark this answers is not in the document]
Comment (GuyAfik) [reply docker/py3-primary/Dockerfile]: done

**Comment by GuyAfik** [comment]
> Looks great! Can we store the binary files somewhere outside the source tree?

removed the binary files, installing libreoffice with wget now

File: docker/build_docker.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (GuyAfik) [reply docker/build_docker.sh:179]: @tkatzir @JudahSchwartz please refer this in py3-tools PR as well - https://github.com/demisto/dockerfiles/pull/9337

File: docker/py3-native/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,103 @@
+FROM devdemisto/py3-tools-ubi:1.0.0.37715
+
+COPY requirements.txt .
+COPY chromium/download_chromedriver.sh .
+COPY repos /repos
+
+# handles chromium + tesseract
```
Comment (tkatzir) [inline docker/py3-native/Dockerfile:7]: Are all these comments needed?

**Review by tkatzir — COMMENTED** [review]
Looks great, one question.

File: docker/build_docker.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (JudahSchwartz) [reply docker/build_docker.sh:179]: @GuyAfik yes

[the remark this answers is not in the document]
Comment (GuyAfik) [reply docker/build_docker.sh:179]: https://github.com/demisto/dockerfiles/pull/10296

File: docker/build_docker.sh [hunk 16]
```diff
@@ -170,6 +170,23 @@ function docker_build {
...
+      echo "starting to install dependencies in poetry..."
+      poetry --version
+      poetry self add poetry-plugin-export
```
Comment (JudahSchwartz) [inline docker/build_docker.sh:182]: move this to dockerbuild also.
We should probably run `poetry env remove` here like we do for pipenv

File: .github/dependabot.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (JudahSchwartz) [reply .github/dependabot.yml:792]: yeah I think it should be pip for poetry, acc to their docs

File: docker/build_docker.sh (hunk 16 above)
Comment (GuyAfik) [reply docker/build_docker.sh:182]: done, once https://github.com/demisto/dockerfiles/pull/10296 and https://github.com/demisto/dockerfiles/pull/9337 are merged ill remove this part.

File: docker/py3-native/Dockerfile (hunk 15 above)
Comment (GuyAfik) [reply docker/py3-native/Dockerfile:7]: no they can be removed before merge, these are just for making things easier for me when to follow up.

File: docker/py3-native/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+# joe-security os dependancies: python3-devel gcc gcc-c++ make wget git
+# slack os dependancies: python3-devel gcc gcc-c++ make wget git libffi
+# office-utils os dependancies: LibreOffice_Linux_x86-64_rpm java-11-openjdk-headless cairo libSM libX11-xcb
+
+
+# line 7 - since in ubi its 7za --> need to make it to run as 7z like our script
+# lines 9-11 is the configuration for sklearn
+# linex 14-16 - install LibreOffice for the office-utils image
+# lines 21-23 - installing libreoffice for the office-utils image
+RUN dnf install -y --nodocs python3-devel gcc gcc-c++ make wget git poppler poppler-utils \
```
Comment (tkatzir) [inline docker/py3-native/Dockerfile:80]: Any reason why this is not part of the previous `RUN` command (that started on line 9) ?

Comment (GuyAfik) [reply docker/py3-native/Dockerfile:80]: no particular reason, it just to differentiate between chromium + tesseract which are "special" images and more "complex" images than others. 

The tesseract and chromium require more OS dependancies than "regular" docker images. 

I think splitting "complex" docker images (which require a lot of os dependancies) - will make it easier to read in the future. 

Shoving all in one RUN command would make it probably less readable

File: docker/py3-native/verify.py [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+
+# sklearn
+import sklearn
+import pandas
+import numpy
+import nltk
+import dill
+import eli5
+import networkx
+print('sklearn is good!!!')
```
Comment (JudahSchwartz) [inline docker/py3-native/verify.py:64]: can we import these tests from where they sit in the repo instead of copying them to here?

File: docker/py3-native/Dockerfile [hunk 19]
```diff
@@ -0,0 +1,103 @@
+FROM demisto/py3-tools-ubi:1.0.0.40055
```
Comment (JudahSchwartz) [inline docker/py3-native/Dockerfile:1]: please add the last updated line like we have in other dockerfiles

File: docker/py3-native/Dockerfile (hunk 17 above)
Comment (tkatzir) [reply docker/py3-native/Dockerfile:80]: Each RUN command adds a layer -> Larger image size.

Comment (GuyAfik) [reply docker/py3-native/Dockerfile:57]: I know, but for server it does not matter that much, also, the differences aren't so extreme.

File: docker/py3-native/Dockerfile (hunk 19 above)
Comment (GuyAfik) [reply docker/py3-native/Dockerfile:1]: we don't have this in all other docker files, why do we need it? git can tell you the last time it was changed

Comment (JudahSchwartz) [reply docker/py3-native/Dockerfile:1]: iits useful for when we want to regenerate the image but we dont have any real changes to make to the file. The build wont catch it without some change

Comment (GuyAfik) [reply docker/py3-native/Dockerfile:1]: do i need to add it manually?

Comment (JudahSchwartz) [reply docker/py3-native/Dockerfile:1]: think youre right its probably not that important

File: docker/py3-native/verify.py (hunk 18 above)
Comment (GuyAfik) [reply docker/py3-native/verify.py:64]: That is a bit problematic, what if someone adds a new dependency to image X, but in the native image we still do not support it? the `verify.py` will fail. If you think that this is good (and then we will know that this dependency needs to be added) i can do it with imports

Comment (GuyAfik) [reply docker/py3-native/verify.py:64]: also, i will need to copy the `verify.py` files to the container, or make it with a symbolic link, however symbolic link would work only a single file and not on all of them. What you want to achieve here is a bit complicated.

File: docker/py3-native/Dockerfile (hunk 17 above)
Comment (GuyAfik) [reply docker/py3-native/Dockerfile:57]: does it matter so much that we will have 2 RUN commands instead of one? do you think it would make the image significantly smaller image?

File: docker/py3-native/verify.py (hunk 18 above)
Comment (JudahSchwartz) [reply docker/py3-native/verify.py:64]: In the case you described, this either will be a reminder to update the master image as well, or we should drop support for that image if the tests dont pass and we dont want to support it
Cant we copy the files in in the same place we do verify.py?
@tkatzir  what do you think?

If this is too high effort for now lets open a task to do it after

Comment (JudahSchwartz) [reply docker/py3-native/verify.py:64]: we will need to touch build code so the docker has the files

File: docker/py3-native/Dockerfile (hunk 17 above)
Comment (tkatzir) [reply docker/py3-native/Dockerfile:57]: You can try.
We can also try and squash layers during the build.

File: docker/py3-native/verify.py (hunk 18 above)
Comment (GuyAfik) [reply docker/py3-native/verify.py]: I think we can do it with symlinks - testing it now. see the latest commit, if that works we can do a symbolic link to each docker image we would like.

File: docker/py3-native/Dockerfile (hunk 17 above)
Comment (GuyAfik) [reply docker/py3-native/Dockerfile:57]: we try to do it in the next version, i don't think its urgent for the upcoming version, it would require me to test everything out again to see i didn't break something

File: docker/py3-native/verify.py (hunk 18 above)
Comment (GuyAfik) [reply docker/py3-native/verify.py]: looks like it works, i updated the docker build a bit to run multiple `verify.py` files

File: docker/build_docker.sh [hunk 20]
```diff
@@ -233,11 +233,13 @@ function docker_build {
         fi
         $PY3CMD ${DOCKER_SRC_DIR}/verify_licenses.py ${image_full_name}
     fi
-    if [ -f "verify.py" ]; then
-        echo "==========================="            
-        echo "Verifying docker image by running the python script verify.py within the docker image"
-        cat verify.py | docker run --rm -i ${image_full_name} python '-'
-    fi
+
+    for filename in `find . -name "*verify.py"`; do
```
Comment (JudahSchwartz) [inline docker/build_docker.sh:237]: We're finding every verify py in the repo?

Comment (GuyAfik) [reply docker/build_docker.sh:237]: no, we are finding every `verify.py` for the entire folder which the docker resides in. in our case it will find all the files that their name is anything_verify.py and that are inside py3-native folder and will run them on the container.

**Comment by GuyAfik** [comment]
@tkatzir @JudahSchwartz please note I have added a short `README.md` as well.

File: docker/py3-primary/verify.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (JudahSchwartz) [reply docker/py3-primary/verify.py]: we implemented this in the end using symlinks

**Review by JudahSchwartz — APPROVED** [review]
What a beast

**Comment by JudahSchwartz** [comment]
lets make sure we get an approval from @glicht before merge

File: docker/py3-native/Dockerfile [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+    && dnf install -y --nodocs --enablerepo=rocky8-base --enablerepo=rocky8-release --enablerepo=fedora-epel ImageMagick \
+    && ./download_chromedriver.sh \
+    && unzip chromedriver_linux64.zip \
+    && mv chromedriver /usr/bin/chromedriver \
+    && rm -rf chromedriver_linux64.zip \
+    && google-chrome --version \
+    && chromedriver --version \
+    && dnf install -y --nodocs git automake make autoconf libtool clang zlib zlib-devel libjpeg libjpeg-devel libwebp libwebp-devel libtiff libtiff-devel libpng libpng-devel pango giflib \
+    && dnf install -y https://vault.centos.org/centos/8/PowerTools/x86_64/os/Packages/giflib-devel-5.1.4-3.el8.x86_64.rpm \
+    && git clone --depth 1 https://github.com/DanBloomberg/leptonica.git \
```
Comment (glicht) [inline docker/py3-native/Dockerfile:28]: I think we should use a tag here to install and not from master. Probably use the latest tag of 1.82.0

Comment (GuyAfik) [reply docker/py3-native/Dockerfile:28]: agreed, will update it tomorrow

File: docker/py3-native/Dockerfile [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+    && cd leptonica \
+    && ./autogen.sh && ./configure && make && make install && ldconfig \
+    && cd .. \
+    && git clone --depth 1 https://github.com/tesseract-ocr/tesseract.git \
```
[the remark this answers is not in the document]
Comment (GuyAfik) [reply docker/py3-native/Dockerfile:32]: agreed, will update it tomorrow
